src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  to the boundary,
+* group instance `group (π_(n+1) x)`,
+* commutative group instance `comm_group (π_(n+2) x)`.
 
-TODO: show that `π n x` is a group when `n > 0` and abelian when `n > 1`. Show that
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO: Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
+  Examples with 𝕊^n (π_n (𝕊^n) = ℤ, π_m (𝕊^n) trivial for m < n).
+  Actions of π_1 on π_n.
+  Group (up to homotopy) of Ω.
+  Lie algebra: ⁅π_(n+1), π_(m+1)⁆ contained in π_(n+m+1).
 
 -/
 
 open_locale unit_interval topology
+open homeomorph
 
 noncomputable theory
 
 universes u
 variables {X : Type u} [topological_space X]
-variables {n : ℕ} {x : X}
+variables {N : Type*} {x : X}
 
-/--
-The `n`-dimensional cube.
--/
+/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
+  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
+  convenient than using `fin (n-1)`. -/
 @[derive [has_zero, has_one, topological_space]]
-def cube (n : ℕ) : Type := fin n → I
-local notation `I^` := cube
+def cube (N : Type*) := N → I
+local notation `I^` n := cube (fin n)
 
 namespace cube
 
-@[continuity] lemma proj_continuous (i : fin n) : continuous (λ f : I^n, f i) :=
-continuous_apply i
-
-/--
-The points of the `n`-dimensional cube with at least one projection equal to 0 or 1.
--/
-def boundary (n : ℕ) : set (I^n) := {y | ∃ i, y i = 0 ∨ y i = 1}
+instance compact_space : compact_space (cube N) :=
+by { convert pi.compact_space, intro, apply_instance }
 
-/--
-The first projection of a positive-dimensional cube.
--/
-@[simp] def head {n} : I^(n+1) → I := λ c, c 0
+instance locally_compact_space : locally_compact_space (cube N) :=
+by convert locally_compact_space.pi; intro; apply_instance
 
-@[continuity] lemma head.continuous {n} : continuous (@head n) := proj_continuous 0
+/-- The points in a cube with at least one projection equal to 0 or 1. -/
+def boundary (N) : set (cube N) := {y | ∃ i, y i = 0 ∨ y i = 1}
 
-/--
-The projection to the last `n` coordinates from an `n+1` dimensional cube.
--/
-@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
+variable {n : ℕ}
+/-- The first projection of a positive-dimensional cube. -/
+@[simps] def head : C(I^(n+1), I) := ⟨λ t, t 0, continuous_apply 0⟩
 
 instance unique_cube0 : unique (I^0) := pi.unique_of_is_empty _
 
-lemma one_char (f : I^1) : f = λ _, f 0 := by convert eq_const_of_unique f
+lemma one_char (f : I^1) : f = λ _, f 0 := eq_const_of_unique f
+
+section
+variable [decidable_eq N]
+
+/-- The forward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation split_at (i : N) := fun_split_at I i
+
+/-- The backward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation insert_at (i : N) := (fun_split_at I i).symm
```

Comment:
I would recommend using `@[reducible] def` instead of `abbreviation`, as the former does not generate equation lemmas, leading to worse docs.

(throughout)